<a href="https://colab.research.google.com/github/adhibgs/data-science-2025/blob/main/Pertemuan10_Adhi_Bagus_May_Maulana_250401020116.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Pertemuan ke-10

*   Nama : Adhi Bagus May Maulana
*   NIM : 250401020116
*   Kelas : IF405

Langkah 1: Muat dan Eksplorasi Data


In [13]:
import pandas as pd
df = pd.read_csv("/content/sample_data/Telco-Customer-Churn.csv")
print(df.shape)
print(df["Churn"].value_counts(normalize=True))

(7043, 21)
Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64


In [14]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Langkah 2: Preprocessing


In [15]:
from sklearn.model_selection import train_test_split

# Mengubah nilai 'Yes'/'No' pada kolom 'Churn' menjadi 1/0
df['Churn'] = df['Churn'].apply(lambda x: 1 if x == 'Yes' else 0)

# Menghilangkan kolom 'customerID' karena tidak relevan sebagai fitur
df_processed = df.drop('customerID', axis=1)

# Melakukan encoding fitur kategorikal menggunakan pd.get_dummies
df_processed = pd.get_dummies(df_processed, drop_first=True)

# Memisahkan X (fitur) dan y (target = Churn)
X = df_processed.drop('Churn', axis=1)
y = df_processed['Churn']

X_tr, X_te, y_tr, y_te = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42)

Langkah 3: Latih Model


In [16]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=300, class_weight="balanced", random_state=42)
rf.fit(X_tr, y_tr)

RandomForestClassifier(class_weight='balanced', n_estimators=300,
                       random_state=42)

Langkah 4: Evaluasi


In [17]:
from sklearn.metrics import classification_report, roc_auc_score
y_pred_rf = rf.predict(X_te)
y_proba_rf = rf.predict_proba(X_te)[:, 1]
print("=== Classification Report ===")
print(classification_report(y_te, y_pred_rf))
print(f"ROC-AUC Score: {roc_auc_score(y_te, y_proba_rf):.4f}")
# TODO: hitung prediksi dan probabilitas
# TODO: tampilkan classification_report dan ROC-AUC


=== Classification Report ===
              precision    recall  f1-score   support

           0       0.83      0.90      0.86      1035
           1       0.64      0.50      0.56       374

    accuracy                           0.79      1409
   macro avg       0.74      0.70      0.71      1409
weighted avg       0.78      0.79      0.78      1409

ROC-AUC Score: 0.8301


Langkah 5: Prediksi Probabilitas dan Simpulkan


In [18]:
sample_proba = pd.DataFrame({
    'Status Churn Aktual': y_te[:5].values,
    'Prediksi Churn': y_pred_rf[:5],
    'Probabilitas Churn (%)': (y_proba_rf[:5] * 100).round(2)
})
print("\n=== Contoh Prediksi Probabilitas Churn ===")
print(sample_proba)
# TODO: hitung probabilitas churn (predict_proba)
# TODO: tulis kesimpulan singkat (3-4 kalimat)


=== Contoh Prediksi Probabilitas Churn ===
   Status Churn Aktual  Prediksi Churn  Probabilitas Churn (%)
0                    0               0                    0.67
1                    0               1                   65.67
2                    0               0                    5.00
3                    0               0                   30.00
4                    0               0                    2.00


Dari Praktikum 10 ini dapat diambil kesimpulan Praktikum ini berhasil menerapkan algoritma Random Forest dengan class_weight="balanced" untuk menangani ketidakseimbangan data target Churn (73.5% No vs 26.5% Yes) dan Model dievaluasi menggunakan Classification Report dan skor ROC-AUC, serta menghasilkan nilai probabilitas (predict_proba) untuk mengukur seberapa besar risiko seorang pelanggan akan mengalami churn